pkg/collector/corechecks/servicediscovery/impl_linux.go

```diff
@@ -0,0 +1,312 @@
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux
+
+package servicediscovery
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/prometheus/procfs"
+
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/servicediscovery/portlist"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+//go:generate mockgen -source=$GOFILE -package=$GOPACKAGE -destination=impl_linux_mock.go
+
+func init() {
+	newOSImpl = newLinuxImpl
+}
+
+var ignoreCfgLinux = []string{
+	"sshd",
+	"dhclient",
+}
+
+type linuxImpl struct {
+	procfs     procFS
+	portPoller portPoller
+	time       timer
+
+	serviceDetector *serviceDetector
+	sender          *telemetrySender
+	ignoreCfg       map[string]bool
+
+	ignoreProcs       map[int]bool
+	aliveServices     map[int]*serviceInfo
+	potentialServices map[int]*serviceInfo
+}
+
+func newLinuxImpl(sender *telemetrySender, ignoreCfg map[string]bool) (osImpl, error) {
+	for _, i := range ignoreCfgLinux {
+		ignoreCfg[i] = true
+	}
+	pfs, err := procfs.NewDefaultFS()
+	if err != nil {
+		return nil, err
+	}
+	poller, err := portlist.NewPoller()
+	if err != nil {
+		return nil, err
+	}
+	return &linuxImpl{
+		procfs:            wProcFS{pfs},
+		portPoller:        poller,
+		time:              realTime{},
+		sender:            sender,
+		serviceDetector:   newServiceDetector(),
+		ignoreCfg:         ignoreCfg,
+		ignoreProcs:       make(map[int]bool),
+		aliveServices:     make(map[int]*serviceInfo),
+		potentialServices: make(map[int]*serviceInfo),
+	}, nil
+}
+
+type processEvents struct {
+	started []*serviceInfo
+	stopped []*serviceInfo
+}
+
+type eventsByName map[string]*processEvents
+
+func (e eventsByName) getOrCreate(name string) *processEvents {
+	events, ok := e[name]
+	if ok {
+		return events
+	}
+	e[name] = &processEvents{}
+	return e[name]
+}
+
+func (e eventsByName) addStarted(svc *serviceInfo) {
+	events := e.getOrCreate(svc.meta.Name)
+	events.started = append(events.started, svc)
+}
+
+func (e eventsByName) addStopped(svc *serviceInfo) {
+	events := e.getOrCreate(svc.meta.Name)
+	events.stopped = append(events.stopped, svc)
+}
+
+func (li *linuxImpl) DiscoverServices() error {
+	procs, err := li.aliveProcs()
+	if err != nil {
+		return fmt.Errorf("failed to get alive processes: %w", err)
+	}
+
+	ports, err := li.openPorts()
+	if err != nil {
+		return fmt.Errorf("failed to get open ports: %w", err)
+	}
+
+	log.Debugf("aliveProcs: %d | ignoreProcs: %d | runningServices: %d | potentials: %d | openPorts: %+v",
+		len(procs),
+		len(li.ignoreProcs),
+		len(li.aliveServices),
+		len(li.potentialServices),
+		ports,
+	)
+
+	var (
+		started []*serviceInfo
+		stopped []*serviceInfo
+	)
+
+	// potentialServices contains processes that we scanned in the previous iteration and had open ports.
+	// we check if they are still alive in this iteration, and if so, we send a start-service telemetry event.
+	for pid, _ := range li.potentialServices {
+		if p, ok := procs[pid]; ok {
+			svc, err := li.getServiceInfo(p, ports)
```

Comment:
since we got the service info when we detected the potential service, there shouldn't be a need to get it again.